capsules/src/public_key_crypto/rsa_keys.rs

```diff
@@ -0,0 +1,154 @@
+//! Helper library for RSA public and private keys
+
+use core::cell::Cell;
+use kernel::hil::entropy;
+use kernel::hil::public_key_crypto::keys::{
+    PubKey, PubPrivKey, PubPrivKeyGenerate, PubPrivKeyGenerateClient, RsaKey, RsaPrivKey,
+};
+use kernel::utilities::cells::OptionalCell;
+use kernel::ErrorCode;
+
+/// A Public/Private RSA 2048 key pair
+pub struct RSA2048Keys<'a> {
+    modulus: [u8; 256],          // Also called n
+    public_exponent: u32,        // Also called e
+    private_exponent: [u8; 256], // Also called d
+    client: OptionalCell<&'a dyn PubPrivKeyGenerateClient<'a>>,
+    pub_initialised: Cell<bool>,
+    priv_initialised: Cell<bool>,
+}
+
+impl<'a> PubPrivKeyGenerateClient<'a> for RSA2048Keys<'a> {
+    fn generation_complete(&'a self, result: Result<(), ErrorCode>) {
+        if result.is_ok() {
+            self.pub_initialised.set(true);
+            self.priv_initialised.set(true);
+        }
+    }
+}
+
+impl<'a> RSA2048Keys<'a> {
+    pub const fn new() -> Self {
+        Self {
+            modulus: [0; 256],
+            // Use the same default as OpenSSL
+            public_exponent: 0x10001,
+            private_exponent: [0; 256],
+            client: OptionalCell::empty(),
+            pub_initialised: Cell::new(false),
+            priv_initialised: Cell::new(false),
+        }
+    }
+}
+
+impl<'a> PubPrivKeyGenerate<'a> for RSA2048Keys<'a> {
+    fn set_client(&'a self, client: &'a dyn PubPrivKeyGenerateClient<'a>) {
+        self.client.set(client);
+    }
+
+    fn generate(
+        &'a self,
+        _trng: &'a dyn entropy::Entropy32,
+        _length: usize,
+    ) -> Result<(), ErrorCode> {
+        unimplemented!();
```

Comment:
Needs to be implemented. I think if it were the inconsistency between the length parameter here and the static lengths returned by `len` would be clear.